lib/client/conntest/kube.go

```diff
@@ -0,0 +1,279 @@
+//   - trace of whether the User role has access to the desired kubernetes cluster: `kubernetes_labels` allow access.
+//   - trace of weather the cluster is accessible and we can list pods on the desired namespace.
+func (s *KubeConnectionTester) TestConnection(ctx context.Context, req TestConnectionRequest) (types.ConnectionDiagnostic, error) {
+	if req.ResourceKind != types.KindKubernetesCluster {
+		return nil, trace.BadParameter("invalid value for ResourceKind, expected %q got %q", types.KindKubernetesCluster, req.ResourceKind)
+	}
+
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(connectionDiagnosticID, map[string]string{},
+		types.ConnectionDiagnosticSpecV1{
+			// We start with a failed state so that we don't need to set it to each return statement once an error is returned.
+			// if the test reaches the end, we force the test to be a success by calling
+			// 	connectionDiagnostic.SetMessage(types.DiagnosticMessageSuccess)
+			//	connectionDiagnostic.SetSuccess(true)
+			Message: types.DiagnosticMessageFailed,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.cfg.UserClient.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	currentUser, err := s.cfg.UserClient.GetCurrentUser(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key, err := client.GenerateRSAKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	certs, err := s.cfg.UserClient.GenerateUserCerts(ctx, proto.UserCertsRequest{
+		PublicKey:              key.MarshalSSHPublicKey(),
+		Username:               currentUser.GetName(),
+		Expires:                time.Now().Add(time.Minute).UTC(),
+		ConnectionDiagnosticID: connectionDiagnosticID,
+		KubernetesCluster:      req.ResourceName,
+	})
+
+	diag, diagErr := s.handleUserGenCertsErr(ctx, connectionDiagnosticID, err)
+	if err != nil || diagErr != nil {
+		return diag, diagErr
+	}
+
+	key.TLSCert = certs.TLS
+
+	certAuths, err := s.cfg.UserClient.GetCertAuthorities(ctx, types.HostCA, false /* loadKeys */)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.TrustedCA = auth.AuthoritiesToTrustedCerts(certAuths)
+	ca, err := s.cfg.UserClient.GetClusterCACert(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ctxWithTimeout, cancelFunc := context.WithTimeout(ctx, req.DialTimeout)
+	defer cancelFunc()
+
+	restConfig := &rest.Config{
+		Host: "https://" + s.cfg.KubernetesPublicProxyAddr,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData:   ca.TLSCA,
+			CertData: key.TLSCert,
+			KeyData:  key.PrivateKeyPEM(),
+		},
+	}
+
+	if s.cfg.TLSRoutingEnabled {
+		k8host := strings.Split(s.webProxyAddr, ":")[0]
+		// replace localhost with "127.0.0.1" so GetKubeTLSServerName can generate a domain kube.cluster.local.
+		if strings.EqualFold(k8host, "localhost") {
+			k8host = "127.0.0.1"
+		}
```

Comment:
This is a little confusing tbh. Is this required to avoid x509 errors in case proxy's public address is `localhost:xxx`? Shouldn't this logic be part of GetKubeTLSServerName then?